### PR TITLE
Fix exception handling on repository

### PR DESCRIPTION
**Pull request by kaizimmerm**
Fixed several issues and worked a bit on API documentation:

* Fixed issue where repository on update operations does not throw ConstraintViolation (as on create) if bean violates a constraint after update.
* Fixed issue where assignDs operations do not throw spring exceptions as expected but native EclipseLink or MySQL exceptions.
* Fixed deadlock on JDBC connections in high load scenarios.
* Added constraint documentation on builders for target and added a few more test as well (should be done for the other entities in similar fashion later on).

Piggy backed a small changed on mgmt API into the request: add "modules" link to DS resource.

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.jdbc.support.SQLStateSQLExceptionTranslator;
+import org.springframework.orm.jpa.JpaSystemException;
+import org.springframework.orm.jpa.vendor.EclipseLinkJpaDialect;
+
+/**
+ * {@link EclipseLinkJpaDialect} with additional exception translation
+ * mechanisms.
+ *
+ */
+public class HawkBitEclipseLinkJpaDialect extends EclipseLinkJpaDialect {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java]: Maybe a few tests would be useful for the corner cases

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    private static final long serialVersionUID = 1L;
+
+    private static final SQLStateSQLExceptionTranslator SQLSTATE_EXCEPTION_TRANSLATOR = new SQLStateSQLExceptionTranslator();
+
+    @Override
+    public DataAccessException translateExceptionIfPossible(final RuntimeException ex) {
+        final DataAccessException dataAccessException = super.translateExceptionIfPossible(ex);
+
+        if (dataAccessException != null) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java]: Avoid negation for better readability.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    private static DataAccessException translateJpaSystemExceptionIfPossible(
+            final DataAccessException accessException) {
+        if (!(accessException instanceof JpaSystemException)) {
+            return accessException;
+        }
+
+        final DataAccessException sql = searchAndTranslateSqlException(accessException);
+        if (sql != null) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java]: Avoid negation for better readability.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 4]
```diff
@@ -214,10 +210,10 @@ public Target findOrRegisterTargetIfItDoesNotexist(final String controllerId, fi
         final JpaTarget target = targetRepository.findOne(spec);
 
         if (target == null) {
-            final Target result = targetManagement.createTarget(entityFactory.target().create()
+            final Target result = targetRepository.save((JpaTarget) entityFactory.target().create()
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: Why do use the repo directly here?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java [hunk 5]
```diff
@@ -84,88 +74,50 @@
...
-        }
+        Exception mappingException = ex;
 
-        Exception mappingException = translatedAccessException;
-
-        LOG.trace("translated excpetion is", translatedAccessException);
         for (final Class<?> mappedEx : MAPPED_EXCEPTION_ORDER) {
 
-            if (mappedEx.isAssignableFrom(translatedAccessException.getClass())) {
+            if (mappedEx.isAssignableFrom(ex.getClass())) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java]: Would prefer early break ;)

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java (hunk 4 above)
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: To avoid that one management proxy is calling another one (wrapped retryAble inside a transaction).

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java [hunk 6]
```diff
@@ -84,88 +74,50 @@
...
-            return accessException;
-        }
-
-        return sqlStateExceptionTranslator.translate(null, null, ex);
-    }
-
-    private static SQLException findSqlException(final JpaSystemException jpaSystemException) {
-        Throwable exception = jpaSystemException.getCause();
-        while (exception != null) {
+    private static Exception findConstraintViolationException(final Exception rex) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java]: Why do you not return `javax.validation.ConstraintViolationException`?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java [hunk 7]
```diff
@@ -84,88 +74,50 @@
...
-                return (SQLException) cause;
+            if (cause instanceof javax.validation.ConstraintViolationException) {
+                return (Exception) cause;
             }
             exception = cause;
-        }
-        return null;
+        } while (exception != null);
+
+        return rex;
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java]: Is this intended not return a null value? So the `TransactionSystemException` is always thrown and will not translated.

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/TargetManagementTest.java [hunk 8]
```diff
@@ -204,50 +189,125 @@ public void createTargetThatViolatesUniqueConstraintFails() {
...
+                        .description(RandomStringUtils.randomAlphanumeric(513))))
+                .as("target with too long description should not be created");
 
-        try {
-            targetManagement.createTarget(entityFactory.target().create().controllerId("     "));
-            fail("target with whitespaces in controller id should not be created");
-        } catch (final ConstraintViolationException e) {
-            // ok
-        }
+        assertThatExceptionOfType(ConstraintViolationException.class)
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/TargetManagementTest.java]: 👍

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java (hunk 1 above)
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/HawkBitEclipseLinkJpaDialect.java]: Javadoc about the concrete exception handling would be nice.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/RepositoryApplicationConfiguration.java [hunk 9]
```diff
@@ -288,7 +289,14 @@ ExceptionMappingAspectHandler createRepositoryExceptionHandlerAdvice() {
 
     @Override
     protected AbstractJpaVendorAdapter createJpaVendorAdapter() {
-        return new EclipseLinkJpaVendorAdapter();
+        return new EclipseLinkJpaVendorAdapter() {
+            HawkBitEclipseLinkJpaDialect jpaDialect = new HawkBitEclipseLinkJpaDialect();
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/RepositoryApplicationConfiguration.java]: `jpaDialect` variable can be final

Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/RepositoryApplicationConfiguration.java]: privat and final

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java (hunk 7 above)
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/aspects/ExceptionMappingAspectHandler.java]: renamed method to make purpose more clear.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java (hunk 4 above)
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: Is this really a issue?
